base-R.qmd

```diff
@@ -0,0 +1,502 @@
+If this pepper shaker is your list `pepper`, then, `pepper[1]` is a pepper shaker containing a single pepper packet, as in @fig-pepper-2.
+`pepper[2]` would look the same, but would contain the second packet.
+`pepper[1:2]` would be a pepper shaker containing two pepper packets.
+`pepper[[1]]` would extract the pepper packet itself, as in @fig-pepper-3.
+
+```{r}
+#| label: fig-pepper-1
+#| echo: false
+#| out-width: "25%"
+#| fig-cap: >
+#|   A pepper shaker that Hadley once found in his hotel room.
+#| fig-alt: >
+#|   A photo of a glass pepper shaker. Instead of the pepper shaker
+#|   containing pepper, it contains many packets of pepper.
+
+knitr::include_graphics("images/pepper.jpg")
+```
+
+```{r}
+#| label: fig-pepper-2
+#| echo: false
+#| out-width: "25%"
+#| fig-cap: >
+#|   `pepper[1]`
+#| fig-alt: >
+#|   A photo of the glass pepper shaker containing just one packet of 
+#|   pepper.
+
+knitr::include_graphics("images/pepper-1.jpg")
+```
+
+```{r}
+#| label: fig-pepper-3
+#| echo: false
+#| out-width: "25%"
+#| fig-cap: >
+#|   `pepper[[1]]`
+#| fig-alt: A photo of single packet of pepper.
+
+knitr::include_graphics("images/pepper-2.jpg")
+```
+
+This same principle applies when you use 1d `[` with a data frame:
+
+```{r}
+df <- tibble(x = 1:3, y = 3:5)
+
+# returns a one-column data frame
+df["x"]
+
+# returns the contents of x
+df[["x"]]
+```
+
+### Exercises
+
+1.  What happens when you subset with a positive integer that's bigger than the length of the vector?
+    What happens when you subset with a name that doesn't exist?
+
+2.  What would `pepper[[1]][1]` be?
+    What about `pepper[[1]][[1]]`?
+
+## Iteration
+
+In @sec-iteration, you learned tidyverse techniques for iteration, primarily `dplyr::across()` and `purrr::map()`.
+In this section, you'll learn about their base equivalents, first the apply family of functions and then the `for` loop, the powerful idea that underlies all these functions.
+
+Some people will tell you to avoid `for` loops because they are slow.
+They're wrong!
+(Well at least they're rather out of date, as `for` loops haven't been slow for many years.) The chief benefit of using functions like `map()` or `apply()` is not speed, but clarity: once you've mastered the basic idea, they make your code faster to write and easier to read.
+
+### Apply family
+
+Another way of saying "map a function over each element of a vector" is "apply a function over each element of a vector", and base R provides a family of functions known as the "apply" family because they all use apply as a suffix.
+Here we'll give you a quick overview of this family so you can recognize them in the wild.
+
+The most important member of this family is `lapply()`, which is very similar to `purrr::map()`[^base-r-3].
+In fact, because we don't use any of `map()`'s more advanced features, you can replace every `map()` call in @sec-iteration with `lapply()`.
```

Comment:
```suggestion
In fact, because we haven't used any of `map()`'s more advanced features, you can replace every `map()` call in @sec-iteration with `lapply()`.
```